Pin docker image to it's checksum

Devils-Knight opened the pull request:
Fixes #187 
Added function to replace `tag` of Docker image with it's `Digest`.

varunsh-coder commented on pinDocker.go:
```diff
@@ -0,0 +1,36 @@
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+)
+
+func pinDocker(action, jobName, inputYaml string) string {
```
@Devils-Knight please create a test file and write test cases. The test cases should cover ghcr.io, docker hub, and may be other providers

Devils-Knight replied:
@varunsh-coder I am unable to find what are the `requests` made by `img, err := remote.Image(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))` this function call.

[Here](https://github.com/google/go-containerregistry/tree/main/pkg/v1/remote) are the details of this function call.

varunsh-coder commented on pinDocker_test.go:
```diff
@@ -0,0 +1,94 @@
+	defer httpmock.DeactivateAndReset()
+
+	// add Table-Driven Tests
+	httpmock.RegisterResponder("GET", "v2/step-security/integration-test/int/manifests/latest",
```
Add https://ghcr.io/ before the v2 part

varunsh-coder commented:
> @varunsh-coder why is this `Get "https://gcr.io/v2/": dial tcp 54.185.253.63:443: connect: connection refused` error poping even after the mock for this is added.
> 
> ```go
> httpmock.RegisterResponder("GET", "https://ghcr.io/v2/",
> 	httpmock.NewStringResponder(200, `{
> 	}`))
> httpmock.RegisterResponder("GET", "https://gcr.io/v2/",
> 	httpmock.NewStringResponder(200, `{
> 	}`))
> ```
> 
> This call is used to ping the docker image. explained [here](https://github.com/google/go-containerregistry/blob/33725d2d7adddb7471eb9f54f2b4e4337f445ccb/pkg/v1/remote/transport/transport.go#L39) [More](https://docs.docker.com/registry/spec/api/#detail) details about this call.

I looked at the calls being intercepted, and none of them were intercepted. 
```
info := httpmock.GetCallCountInfo()
```

Please read this:
https://github.com/jarcoal/httpmock/issues/99
I think you need to set the transport when you call the docker related libraries

Devils-Knight commented:
> 
> I looked at the calls being intercepted, and none of them were intercepted.
> 
> ```
> info := httpmock.GetCallCountInfo()
> ```
> 
> Please read this: [jarcoal/httpmock#99](https://github.com/jarcoal/httpmock/issues/99) I think you need to set the transport when you call the docker related libraries

Thanks @varunsh-coder this helped a lot, now all the calls are being intercepted.

![image](https://user-images.githubusercontent.com/73299843/153060651-84c386d3-fc37-40f8-bd05-8f4df9bdc580.png)

Devils-Knight commented:
@varunsh-coder I don't understand why httpmock is returning different digest value, if I have assigned the right value to digest in the response.
>Wrong Output:
> ghcr.io/step-security/integration-test/int@sha256:c4f6f91aebee0fd34baa44ee23e0a73accf2d9cd3c3e991457bb14f1b41ee4e4
```.go
httpmock.RegisterResponder("GET", "https://ghcr.io/v2/step-security/integration-test/int/manifests/latest",
		func(req *http.Request) (*http.Response, error) {
			resp := httpmock.NewStringResponse(200, `{
				"digest": "sha256:f1f95204dc1f12a41eaf41080185e2d289596b3e7637a8c50a3f6fbe17f99649",
				"mediaType": "application/vnd.docker.distribution.manifest.v2+json",
				"size": 1568,
				"config": {
					"digest": "sha256:d1ea12d5b843717d90bca617e78609dbcc2ebcc9386c9edd98ecdb1427469b43",
					"mediaType": "application/vnd.docker.container.image.v1+json",
					"size": 2868
				}
			}`)
			resp.Header.Add("Docker-Content-Digest", "sha256:f1f95204dc1f12a41eaf41080185e2d289596b3e7637a8c50a3f6fbe17f99649")
			return resp, nil
		},
	)
```
I have added this manifest response acc. to [this](https://docs.docker.com/registry/spec/api/#manifest) format.

The `digest` of the image is added in the header of the response i.e I have added the `function` instead of `httpmock.NewStringResponder`.

Devils-Knight commented:
I think the inconsistency is caused by [this](https://github.com/google/go-containerregistry/blob/33725d2d7adddb7471eb9f54f2b4e4337f445ccb/pkg/v1/remote/descriptor.go#L267) part.
I tried adding `content-type` header to remove inconsistency by fetching `digest` from header.([here](https://github.com/google/go-containerregistry/blob/33725d2d7adddb7471eb9f54f2b4e4337f445ccb/pkg/v1/remote/descriptor.go#L274))
But the error is poping by adding the above content type because of [this](https://github.com/google/go-containerregistry/blob/33725d2d7adddb7471eb9f54f2b4e4337f445ccb/pkg/v1/remote/descriptor.go#L177)

In normal condition the code run fines because it uses [this](https://github.com/google/go-containerregistry/blob/33725d2d7adddb7471eb9f54f2b4e4337f445ccb/pkg/v1/remote/descriptor.go#L267) part of code to calculate hash from body.
In test case we have to add the exact body of response, otherwise a different hash will generate.

So what should I do about this @varunsh-coder sir.

Devils-Knight commented:
@varunsh-coder I have updated the `httpmock.NewStringResponder` to fetch manifest file from `response` folder and also added the steps for the same. Please review the changes sir.

varunsh-coder commented on pindockers_test.go:
```diff
@@ -0,0 +1,70 @@
+	//step.1> Get Token by this api call (https://auth.docker.io/token?service=<service>&scope=repository:<name>:<tag>)
+	//step.2> Get manifest file by setting header as (Authorization : Bearer <token>) and using this api call (https://<service>/v2/<name>/manifests/<tag>)
+	//step.3> Download response and save it in this path (testfiles/pindockers/response/<name>.txt)
+	httpmock.RegisterResponder("GET", "https://ghcr.io/v2/step-security/integration-test/int/manifests/latest",
```
@Devils-Knight please find an example where docker hub is used to pull image. Docker hub is also popular, and if it can be used in such a context, we should support it. I think if the host name of registry is not specified, it might default to docker hub.

varunsh-coder commented on pindockers.go:
```diff
@@ -0,0 +1,62 @@
+
+var Tr http.RoundTripper = remote.DefaultTransport
+
+func PinDockers(inputYaml string) (string, error) {
```
@Devils-Knight can we call this PinDocker (remove the `s`)

varunsh-coder commented on pindockers.go:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	img, err := remote.Image(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain), remote.WithTransport(Tr))
+	if err != nil {
```
@Devils-Knight ideally we should log the error somewhere, else we will not know things are failing. Can you please create an issue to track this?

varunsh-coder commented on testfiles/pindockers/response/gcrResponse.txt:
```diff
@@ -0,0 +1,31 @@
+{
+   "schemaVersion": 2,
```
@Devils-Knight file extension should json.
